src/raft.rs

```diff
@@ -1688,20 +1711,52 @@ impl<T: Storage> Raft<T> {
                 }
 
                 for (i, e) in m.mut_entries().iter_mut().enumerate() {
+                    let mut cc;
                     if e.get_entry_type() == EntryType::EntryConfChange {
-                        if self.has_pending_conf() {
-                            info!(
-                                self.logger,
-                                "propose conf entry ignored since pending unapplied configuration";
-                                "entry" => ?e,
-                                "index" => self.pending_conf_index,
-                                "applied" => self.raft_log.applied,
-                            );
-                            *e = Entry::default();
-                            e.set_entry_type(EntryType::EntryNormal);
+                        let mut cc_v1 = ConfChange::default();
+                        if let Err(e) = cc_v1.merge_from_bytes(e.get_data()) {
+                            error!(self.logger, "invalid confchange"; "error" => ?e);
+                            return Err(Error::ProposalDropped);
+                        }
+                        cc = cc_v1.into_v2();
+                    } else if e.get_entry_type() == EntryType::EntryConfChangeV2 {
+                        cc = ConfChangeV2::default();
+                        if let Err(e) = cc.merge_from_bytes(e.get_data()) {
+                            error!(self.logger, "invalid confchangev2"; "error" => ?e);
+                            return Err(Error::ProposalDropped);
+                        }
+                    } else {
+                        continue;
+                    }
+
+                    let reason = if self.has_pending_conf() {
+                        "possible unapplied conf change"
+                    } else {
+                        let already_joint = confchange::joint(self.prs.conf());
+                        let want_leave = cc.changes.is_empty();
+                        if already_joint && !want_leave {
+                            "must transition out of joint config first"
+                        } else if !already_joint && want_leave {
+                            "not in joint state; refusing empty conf change"
                         } else {
-                            self.pending_conf_index = self.raft_log.last_index() + i as u64 + 1;
+                            ""
                         }
+                    };
+
+                    if reason.is_empty() {
+                        self.pending_conf_index = self.raft_log.last_index() + i as u64 + 1;
+                    } else {
+                        info!(
```

Comment:
Should all the above reasons be set as info level? For example, does empty conf change actually exist?

Reply:
Yes. See https://github.com/tikv/raft-rs/pull/386/files#diff-18f3dab6275bbc16b4a102e73378c08bR828-R832.